File: 53.jump-game/solution.kt
class Solution {
    fun canJump(nums: IntArray): Boolean {
        val N = nums.size
        if (N == 1) return true
        var i = 0
        var maxJmp = 0
        while ((i < N) && (i <= maxJmp) && (maxJmp < N)) {
            maxJmp = maxOf(maxJmp, i + nums[i])
            i++
        }
        return maxJmp >= N - 1
    }
}

File: 53.jump-game/solution.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        N = len(nums)
        if N == 1:
            return True

        maxJmp = 0
        i = 0
        while (i < N) and (i <= maxJmp) and (maxJmp < N):
            maxJmp = max(maxJmp, i + nums[i])
            i += 1
        return maxJmp >= N - 1
